### ISSUE TITLE
gcache.page.xxx no deleting, even after we upgrade to galera version > 3.12 

### ISSUE THREAD

**Issue by Xiaohua-Shen** [percona/galera#91]
we have an galera cluster with 2 node and 1 garbd arbiter, let me named them as node1, node2 and node-garbd.

current installed version:
```
#:/etc/mysql# dpkg -l | grep percona
ii  percona-xtrabackup                       2.3.5-1.trusty                    amd64        Open source backup tool for InnoDB and XtraDB
ii  percona-xtradb-cluster-client-5.6        5.6.32-25.17-1.trusty             amd64        Percona XtraDB Cluster database client binaries
ii  percona-xtradb-cluster-common-5.6        5.6.32-25.17-1.trusty             amd64        Percona XtraDB Cluster database common files (e.g. /etc/mysql/my.cnf)
ii  percona-xtradb-cluster-galera-3.x        3.17-1.trusty                     amd64        Galera components of Percona XtraDB Cluster
ii  percona-xtradb-cluster-server-5.6        5.6.32-25.17-1.trusty             amd64        Percona XtraDB Cluster database server binaries

```
Related OS and kernel: 
```
#:/etc/mysql# uname -a
Linux controller2 3.13.0-100-generic #147-Ubuntu SMP Tue Oct 18 16:48:51 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
#:/etc/mysql# cat /etc/issue
Ubuntu 14.04.5 
```
Related mysql conf:
```
[mysqld]

general-log = 0
general-log-file = /var/log/mysql/mysql.log
log-error = /var/log/mysql/mysql.err
log-warnings = 10
datadir = /var/lib/mysql
user = mysql
bind-address = 0.0.0.0

wsrep_provider = /usr/lib/libgalera_smm.so
wsrep_cluster_address = gcomm://192.168.0.1,192.168.0.2
wsrep_slave_threads = 4

wsrep_sst_method = xtrabackup-v2
wsrep_sst_auth = "sst_admin:xxxxxxxxxxxxxxxxxxxxxxxxxxx"

wsrep_cluster_name = db_cluster
wsrep_node_address = 192.168.0.1

default_storage_engine = InnoDB
binlog_format = ROW
innodb_autoinc_lock_mode = 2
innodb_locks_unsafe_for_binlog = 1
# only keep 2 gcache page storage
wsrep_provider_options="gcache.size=1G; gcache.keep_pages_count=2"

max_connections = 1000
innodb_log_file_size=50331648
lock_wait_timeout = 1800

[mysqld]
collation-server = utf8_unicode_ci
character-set-server = utf8

[client]
default-character-set=utf8

[mysql]
default-character-set=utf8
```

Scenario:

1. node1 is the first member of the cluster, then node2 and node-garbd join.  they are active-active
2. writes always happen at node 1, and node2 replication works well. 

Symptom:

1. node1's gcache.page.xxxx files continuous growing, and no deletion.
```
root@controller1:~# zgrep "gcache" /var/log/mysql/mysql.err.*.gz
2016-11-08 08:17:00 6972 [Note] WSREP: Passing config to GCS: base_dir = /var/lib/mysql/; base_host = 10.140.51.202; base_port = 4567; cert.log_conflicts = no; debug = no; evs.auto_evict = 0; evs.delay_margin = PT1S; evs.delayed_keep_period = PT30S; evs.inactive_check_period = PT0.5S; evs.inactive_timeout = PT15S; evs.join_retrans_period = PT1S; evs.max_install_timeouts = 3; evs.send_window = 4; evs.stats_report_period = PT1M; evs.suspect_timeout = PT5S; evs.user_send_window = 2; evs.view_forget_timeout = PT24H; gcache.dir = /var/lib/mysql/; gcache.keep_pages_count = 2; gcache.keep_pages_size = 0; gcache.mem_size = 0; gcache.name = /var/lib/mysql//galera.cache; gcache.page_size = 128M; gcache.size = 1G; gcomm.thread_prio = ; gcs.fc_debug = 0; gcs.fc_factor = 1.0; gcs.fc_limit = 16; gcs.fc_master_slave = no; gcs.max_packet_size = 64500; gcs.max_throttle = 0.25; gcs.recv_q_hard_limit = 9223372036854775807; gcs.recv_q_soft_limit = 0.25; gcs.sync_donor = no; gmcast.segment = 0; gmcast.version = 0; pc.announce_timeout = PT3S; pc.checksum = fals
2016-11-09 07:47:56 6972 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000000 of size 134217728 bytes
2016-11-09 09:55:52 6972 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000001 of size 134217728 bytes
2016-11-09 11:52:45 6972 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000002 of size 134217728 bytes
2016-11-09 13:53:48 6972 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000003 of size 134217728 bytes
2016-11-09 15:54:50 6972 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000004 of size 134217728 bytes
2016-11-09 17:55:02 6972 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000005 of size 134217728 bytes
2016-11-09 19:55:02 6972 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000006 of size 134217728 bytes
2016-11-09 21:55:01 6972 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000007 of size 134217728 bytes
2016-11-09 23:55:02 6972 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000008 of size 134217728 bytes
2016-11-10 01:55:02 6972 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000009 of size 134217728 bytes
2016-11-10 03:55:01 6972 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000010 of size 134217728 bytes

```
2. node2's gcache.page.xxxx files can be deleted as the configuration (keep 2 pages only)
```
#:/var/lib/mysql# zgrep "gcache" /var/log/mysql/mysql.err.*.gz
2016-11-08 08:17:53 6639 [Note] WSREP: Passing config to GCS: base_dir = /var/lib/mysql/; base_host = 10.140.51.203; base_port = 4567; cert.log_conflicts = no; debug = no; evs.auto_evict = 0; evs.delay_margin = PT1S; evs.delayed_keep_period = PT30S; evs.inactive_check_period = PT0.5S; evs.inactive_timeout = PT15S; evs.join_retrans_period = PT1S; evs.max_install_timeouts = 3; evs.send_window = 4; evs.stats_report_period = PT1M; evs.suspect_timeout = PT5S; evs.user_send_window = 2; evs.view_forget_timeout = PT24H; gcache.dir = /var/lib/mysql/; gcache.keep_pages_count = 2; gcache.keep_pages_size = 0; gcache.mem_size = 0; gcache.name = /var/lib/mysql//galera.cache; gcache.page_size = 128M; gcache.size = 1G; gcomm.thread_prio = ; gcs.fc_debug = 0; gcs.fc_factor = 1.0; gcs.fc_limit = 16; gcs.fc_master_slave = no; gcs.max_packet_size = 64500; gcs.max_throttle = 0.25; gcs.recv_q_hard_limit = 9223372036854775807; gcs.recv_q_soft_limit = 0.25; gcs.sync_donor = no; gmcast.segment = 0; gmcast.version = 0; pc.announce_timeout = PT3S; pc.checksum = fals
2016-11-09 07:47:56 6639 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000000 of size 134217728 bytes
2016-11-09 09:55:52 6639 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000001 of size 134217728 bytes
2016-11-09 11:52:45 6639 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000002 of size 134217728 bytes
2016-11-09 11:52:45 6639 [Note] WSREP: Deleted page /var/lib/mysql/gcache.page.000000
2016-11-09 13:53:48 6639 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000003 of size 134217728 bytes
2016-11-09 13:53:48 6639 [Note] WSREP: Deleted page /var/lib/mysql/gcache.page.000001
2016-11-09 15:54:50 6639 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000004 of size 134217728 bytes
2016-11-09 15:54:50 6639 [Note] WSREP: Deleted page /var/lib/mysql/gcache.page.000002
2016-11-09 17:55:02 6639 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000005 of size 134217728 bytes
2016-11-09 17:55:02 6639 [Note] WSREP: Deleted page /var/lib/mysql/gcache.page.000003
2016-11-09 19:55:02 6639 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000006 of size 134217728 bytes
2016-11-09 19:55:02 6639 [Note] WSREP: Deleted page /var/lib/mysql/gcache.page.000004
2016-11-09 21:55:01 6639 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000007 of size 134217728 bytes
2016-11-09 21:55:01 6639 [Note] WSREP: Deleted page /var/lib/mysql/gcache.page.000005
2016-11-09 23:55:02 6639 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000008 of size 134217728 bytes
2016-11-09 23:55:02 6639 [Note] WSREP: Deleted page /var/lib/mysql/gcache.page.000006
2016-11-10 01:55:02 6639 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000009 of size 134217728 bytes
2016-11-10 01:55:02 6639 [Note] WSREP: Deleted page /var/lib/mysql/gcache.page.000007
2016-11-10 03:55:01 6639 [Note] WSREP: Created page /var/lib/mysql/gcache.page.000010 of size 134217728 bytes
2016-11-10 03:55:01 6639 [Note] WSREP: Deleted page /var/lib/mysql/gcache.page.000008
```
3. if I restart node1's mysql, it's log will have error like below:
```
2016-11-10 09:44:50 6972 [ERROR] WSREP: gcache/src/gcache_page_store.cpp:~PageStore():243: Could not delete 11 page files: some buffers are still "mmapped".
```
4. if stop node2's mysql service ,no above error log . but after restart , this node's gcache.page will keep without deleting, just like node1.

We have many envs with same configuration, and they have both symptom.

**Comment by kbauskar** [comment]
* Pages remains pinned if the purge criteria is not yet met.
* Are you using small number of huge-transactions ?

---------

If you have a reproducible test-case that you can share it will help accelerate tracing the issue.

**Comment by Xiaohua-Shen** [comment]
it's used in openstack . openstack related services are the only user for db.
traditional cases are  big number of small-transactions (The token table have frequently R/W).
and some token-expire record cleaning happened every 5 minutes, hundreds of records be deleted in a batch.   

we have one env started running for > 1 month, and after installation, related mysql didn't restart. 
same symptom as I mentioned above:  one node keep only 2 gcache.page files, the other one already have 1129 page files.  and related mysqld process used memory is increasing too, now it already used up 30G mem.

can you tell me the purge criteria?  I already enable wsrep_debug, and can get wsrep related variables/status/logs to you if you instruct me. 

I have no reproducible test-cases now, and will give to you if I have more idea about that.

**Comment by Xiaohua-Shen** [comment]
Also we have huge number of warning in the logs like below, not sure if these two symptoms have linkage.
```
[Warning] Aborted connection 1465541 to db: 'keystone' user: 'keystone' host: 'xxx.xxx.xxx.xxx' (Got an error reading communication packets)
```
when open wsrep_debug, huge number of below logs too, (all the trx number are same):
```
[Note] [Debug] WSREP: galera/src/wsrep_provider.cpp:galera_post_rollback():432: trx 18446744073709551615 not found
```
And wsrep_local_cached_downto always be 1 for the first cluster node before it restart.
```
~# mysql -uroot -e "show status like 'wsrep_%'" | grep -E "wsrep_gcache_pool_size|wsrep_local_cached_downto|wsrep_last_committed|wsrep_local_commit"
wsrep_last_committed    65842641
wsrep_local_commits     1
wsrep_local_cached_downto       1
wsrep_gcache_pool_size  153000734232
```

**Comment by Xiaohua-Shen** [comment]
I already found the cause.  we have some serverspec check for mysql configuration, which caused the gcache issue.
https://github.com/mizzy/serverspec/blob/master/lib/serverspec/type/mysql_config.rb#L4
after we remove related serverspec check, no gcache.page issue now.

**Comment by kbauskar** [comment]
@Xiaohua-Shen 
Not sure If got your last reply. What kind of checks did you have that caused the problem ?

**Comment by rameshvs02** [comment]
This issue is fixed in PXC-5.6.42
For more info please refer : https://jira.percona.com/browse/PXC-2220
